src/engine/AmguConsumer.sol

```diff
@@ -3,49 +3,55 @@ pragma solidity 0.6.1;
 import "../dependencies/DSMath.sol";
 import "../dependencies/token/IERC20.sol";
 import "../prices/IPriceSource.sol";
+import "../version/IRegistry.sol";
 import "../version/IVersion.sol";
 import "./IEngine.sol";
-import "../version/Registry.sol";
 
 /// @notice Abstract contracts
 /// @notice inherit this to pay AMGU on a function call
 abstract contract AmguConsumer is DSMath {
 
-    /// @dev each of these must be implemented by the inheriting contract
-    function engine() public view virtual returns (address);
-    function mlnToken() public view virtual returns (address);
-    function priceSource() public view virtual returns (address);
-    function registry() public view virtual returns (address);
-    event AmguPaid(address indexed payer, uint256 totalAmguPaidInEth, uint256 amguChargableGas, uint256 incentivePaid);
+    event AmguPaid(
+        address indexed payer,
+        uint256 totalAmguPaidInEth,
+        uint256 amguChargableGas,
+        uint256 incentivePaid
+    );
 
-    /// bool deductIncentive is used when sending extra eth beyond amgu
-    modifier amguPayable(bool deductIncentive) {
+    IRegistry public registry;
+
+    constructor (address _registry) public {
+        registry = IRegistry(_registry);
+    }
+
+    /// @param _deductIncentive is used when sending extra eth beyond amgu
+    modifier amguPayable(bool _deductIncentive) {
         uint preGas = gasleft();
         _;
         uint postGas = gasleft();
 
-        uint mlnPerAmgu = IEngine(engine()).getAmguPrice();
+        uint mlnPerAmgu = IEngine(registry.engine()).getAmguPrice();
         uint mlnQuantity = mul(
             mlnPerAmgu,
             sub(preGas, postGas)
         );
-        address nativeAsset = Registry(registry()).nativeAsset();
-        uint ethToPay = IPriceSource(priceSource()).convertQuantity(
+        address nativeAsset = registry.nativeAsset();
+        uint ethToPay = IPriceSource(registry.priceSource()).convertQuantity(
             mlnQuantity,
-            mlnToken(),
+            registry.mlnToken(),
             nativeAsset
         );
         uint incentiveAmount;
-        if (deductIncentive) {
-            incentiveAmount = Registry(registry()).incentive();
+        if (_deductIncentive) {
+            incentiveAmount = registry.incentive();
         } else {
             incentiveAmount = 0;
         }
         require(
             msg.value >= add(ethToPay, incentiveAmount),
             "Insufficent AMGU and/or incentive"
         );
-        IEngine(engine()).payAmguInEther.value(ethToPay)();
+        IEngine(registry.engine()).payAmguInEther.value(ethToPay)();
 
         require(
             msg.sender.send(
```

Comment:
check out Sean's comments (hidden by default now I think). We added the first require so that it would fail earlier potentially. Also for transfer I would agree except the error messages are not the best without embedding them in `require`s. I think we can change all `send` to `transfer` when we upgrade to solidity 0.6.3 though, since those errors should be better